Cleanup of IndexSettingProvider

felixbarny opened the pull request:
In https://github.com/elastic/elasticsearch/pull/133232, we've added the ability to provide index metadata with an IndexSettingProvider. It turned out that we don't need that functionality as we ended up using a private index setting in https://github.com/elastic/elasticsearch/pull/132566.

This also adds the `IndexVersion` as another parameter. This is in preparation for [this](https://github.com/elastic/elasticsearch/pull/132566#discussion_r2366143253) suggestion to conditionally set one or another setting, depending on the index version.

felixbarny committed:
Cleanup of IndexSettingProvider

In https://github.com/elastic/elasticsearch/pull/133232, we've added the ability to provide index metadata with an IndexSettingProvider.
It turned out that we don't need that functionality as we ended up using a private index setting in https://github.com/elastic/elasticsearch/pull/132566.

This also adds the `IndexVersion` as another parameter. This is in preparation for [this](https://github.com/elastic/elasticsearch/pull/132566#discussion_r2366143253) suggestion to conditionally set one or another setting, depending on the index version.

Commit:
[CI] Auto commit changes from spotless

felixbarny committed:
Merge branch 'main' into index-settings-provider-index-version

dakrone reviewed (approved):
LGTM, I left one comment

felixbarny committed:
Use index version of write index

felixbarny committed:
Merge remote-tracking branch 'origin/main' into index-settings-provider-index-version

felixbarny committed:
Fix resolution of IndexVersion

felixbarny committed:
Merge remote-tracking branch 'origin/main' into index-settings-provider-index-version

felixbarny committed:
Merge remote-tracking branch 'felixbarny/index-settings-provider-index-version' into index-settings-provider-index-version

felixbarny committed:
Adjust JavaDoc of IndexSettingProvider

It can't provide custom metadata anymore

felixbarny commented on modules/data-streams/src/main/java/org/elasticsearch/datastreams/action/TransportGetDataStreamsAction.java:
```diff
@@ -199,16 +200,16 @@ static IndexMode resolveMode(
                 List.of(),
-                builder,
-                (k, v) -> {}
+                IndexVersion.current(),
```
Implemented in [7a15254](https://github.com/elastic/elasticsearch/pull/135251/commits/7a15254ae1826d7aa691d4400009e5df72f4ae1a)

felixbarny committed:
Disallow providers to set the index version

Otherwise, they can't rely on the provided version to be the one that will be actually used for the created index.

felixbarny committed:
Merge branch 'main' into index-settings-provider-index-version
